### PR TITLE
New Locomotion System

### PR DESCRIPTION
**Pull request by FejZa**
# XRTK - Mixed Reality Toolkit Pull Request

## Overview

This PR introduces the new locomotion system which replaces the old teleport system. This new system is set up in a way that it is extremely easy for users to implement new types of locomotion if needed but also comes with common default implementations for smooth locomotion and teleportation.

Key Features:
- Change locomotion types at runtime
- Register locomotion implementations as data providers (no MonoBehaviours)
- Very flexible in use and extend

https://user-images.githubusercontent.com/9565734/129541455-ae127fe7-d0a1-4590-b6ce-bbf1f6910485.mp4

## Changes
- Remove Teleport System
- Add Locomotion System

## Breaking Changes
- Teleport System is gone, require updating existing root profiles to register the new Locomotion System

## Related Submodule Changes

- https://github.com/XRTK/SDK/pull/243
- https://github.com/XRTK/Examples/pull/47

### CODE REVIEW

**Comment by StephenHodgson** [comment]
Please double check that all the submodules are pointing to the right commits

**Comment by SimonDarksideJ** [comment]
Curious, the UltraLeap changes in that Branch?  should that be in this PR?

**Comment by FejZa** [comment]
Nope. I gotta do some clean up here, not sure what happened. I'll let you know.

**Comment by SimonDarksideJ** [comment]
Bit curious as to the alignment of implementations in base and what is in the SDK.

I was under the impression we were just going to have a single (default) Locomotion provider in the core and all the others would be in the SDK, am I incorrect?

If so, all inspectors and relevant providers in the SDK should also have their inspectors / profile definitions and so on in the SDK.

Also find it curious the "BaseTeleportProvider" is in the SDK and not core?

**Comment by FejZa** [comment]
We talked about this a while ago. I don't like having to move anything to SDK. All the default implementations are in Core. Especially base classes like `BaseLocomotionProvider` etc. Because if they weren't, then I couldn't make use of them in Core implementations without creating a dependency on SDK. Which does not make sense.

In general, back then when we talked about it, we agreed to move things to Core so they can work on their own without the SDK.

The only thing living in SDK are the default profiles basically.

**Comment by FejZa** [comment]
Oh yeah sorry, Pointers and HotSpot are in SDK as well (which I would love to move to Core tbh) :D

**Comment by StephenHodgson** [comment]
> Pointers and HotSpot

Makes sense to me. 

Yeah the general consensus was to try to keep monobehaviours out of core as much as possible for future engine porting. But at this point I don't think it really matters. I think we should just stick to Unity.

**Comment by FejZa** [comment]
I see. Well I think we can leave it like this for now. And yes I think we should just stick to Unity.

I am really torn in between making locomotion its own package and not.

**Comment by StephenHodgson** [comment]
> I am really torn in between making locomotion its own package and not.

I think it is ubiquitous enough to stay in core

**Comment by FejZa** [comment]
Renamed hotspots to TeleportAnchor

**Comment by StephenHodgson** [comment]
Needs to be fixed. It won't pass until the packages are updated I think

**Comment by FejZa** [comment]
Any idea how I can reset submodules? This PR shows I made changes in Ultraleap and glTF, which does not make any sense. Not sure what happened here.

**Comment by StephenHodgson** [comment]
I can take a quick look

File: XRTK-Core/Assets/Assets.index [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (StephenHodgson) [inline XRTK-Core/Assets/Assets.index:1]: What does this file do?

File: XRTK-Core/Packages/com.xrtk.core/Runtime/EventDatum/Diagnostics/BaseDiagnosticsEventData.cs [hunk 2]
```diff
@@ -5,7 +5,7 @@
 using UnityEngine.EventSystems;
 using XRTK.Extensions;
 
-namespace XRTK.EventDatum.DiagnosticsSystem
+namespace XRTK.EventDatum.Diagnostics
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Runtime/EventDatum/Diagnostics/BaseDiagnosticsEventData.cs:8]: Not sure if we should include this kind of breaking change in with the Locomotion changes.

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs [hunk 3]
```diff
@@ -48,11 +47,6 @@ public interface IMixedRealityPointer : IEqualityComparer
         /// </summary>
         ICursorModifier CursorModifier { get; set; }
 
-        /// <summary>
-        /// The currently active teleport hotspot.
-        /// </summary>
-        IMixedRealityTeleportHotSpot TeleportHotSpot { get; set; }
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: Not replaced with ITeleportAnchor?

Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: Huh? Let me check this I updated ALL of the references and specifically searched for the string "Hotspot" and updated.

File: XRTK-Core/Assets/Assets.index (hunk 1 above)
Comment (FejZa) [reply XRTK-Core/Assets/Assets.index:1]: This file is created by the `Quick Search` upm package. I accidentally checked it in.

**Review by StephenHodgson — CHANGES_REQUESTED** [review]
I think the namespace changes are a bit out of scope for this. 

We can address those in the next release cycle

**Comment by FejZa** [comment]
Thanks. I'll revert some of the changes now.

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs (hunk 3 above)
Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: Ah, I misread this. This is fine. Pointers do no longer have a reference to hotspots/anchors in their interface. So this change is correct.

**Comment by FejZa** [comment]
Feedback addressed @StephenHodgson.

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs (hunk 3 above)
Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: Interesting, how do we know which anchor is active for the specific pointer then?

Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: The active anchor is available via `ITeleportTargetProvider.cs`, which happens to be implemented by the `TeleportPointer`. Generally providing a teleport target is not tied to pointers. Anything implementing the mentioned interface can provide teleport targets.

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: This was a convenience property for those consuming pointer event data, so they knew which specfic anchor the pointer was targeting

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: > The active anchor is available via ITeleportTargetProvider.cs

Sounds an awful lot like that pointer mediator 😂

I'll late a look and get back to you.

Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: It does not. It's simply an interface for components providing a teleport target. You could be throwing a ball and where it lands that would be your teleport target. See how that is not a pointer?

Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: You need to let go of the idea that only pointers provide teleportation targets.

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Interfaces/InputSystem/IMixedRealityPointer.cs:54]: @FejZa That's a fair point 😅

**Comment by StephenHodgson** [comment]
LGTM, hopefully the builds start passing again then we'll merge it in

**Comment by StephenHodgson** [comment]
@FejZa I think I figured it out. I reverted the gltf submodule changes. It seems to hang the editor when importing models for some reason. I'll look into it.

**Review by StephenHodgson — APPROVED** [review]

**Comment by FejZa** [comment]
Great, thanks. Once the package is published I need to bump SDK dependencies so the build can pass.

**Comment by StephenHodgson** [comment]
It's already published

**Pull request merged**